### ISSUE TITLE
Interactive console won't recognize modified code in an imported module

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#62]
_From [4.derek....@gmail.com](https://code.google.com/u/111899204237312999915/) on 2009-10-04T14:17:33Z_

What steps will reproduce the problem?

 1. run code in interactive console, where code calls a function from an
imported module. ...say mymodule.func()
2. mymodule.func() has an error and an exception is raised (as expected)
3. modify code in mymodule.func() to correct the problem (and even save
mymodule.py)
4. interactive console raises the same exception (i.e., ignores the code
modification) on the next execute.
5. nothing short of restarting spyder will get interactive console to
recognize the changes made to the code. What is the expected output? What do you see instead? Expect code changes in the entire stack to be reflected on execute (with or
without a resave).  Note: this behaviour does not occur if the exception
occurs in the main module, nor have I noticed this issue using the external
console. What version of the product are you using? On what operating system? Spyder 1.0beta10 (with XP) and pydee 0.4.13 (XP and windows server 2003) 

Please provide any additional information below

. Tried to look for documentation on this problem with no luck.  Otherwise, a
really great program, keep up the good work!

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=62_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-04T13:03:05Z_

Actually you're pointing out what could be seen as a limitation of the Python
language (which is not a problem in fact because this way of working is on many other
situations a great advantage): the only way to take into account modification in a
function or a class (anything outside the main module as you righteously mentioned)
is to reload the module explicitely -- using the 'reload' function (usage:
reload(mymodule) in your example, before calling the modified version of mymodule.func).
Note that, if you have IPython installed on your machine, another useful function is
available: dreload (for "deep" reload), which will reload recursively all modules
imported by 'mymodule' -- sometimes it won't work well.

Anyway, this is not a bug from Spyder but rather a property of the Python language
that you have to know to understand this behaviour.

Another workaround could be to use the external console instead of the interactive
console, and then restarting the console itself would be sufficient to start from
scratch. But it depends on what you're doing exactly in this console session (the
external console does not allow to use matplotlib with the interactive option turned
on, but apart from this, it's very close to the interactive console).

So, I will try and write something in the documentation to explain all this.
Basically if I explain things as written above, is this answer your questions?

**Summary:** Documentation: explain why interactive console won't recognize modified code in an imported module  
  **Status:** Accepted  
  **Labels:** -Type-Defect Type-Enhancement Milestone-Release1.0

**Comment by spyder-issues-migrator** [comment]
_From [4.derek....@gmail.com](https://code.google.com/u/111899204237312999915/) on 2009-10-04T15:05:48Z_

Thank you Pierre, that helps me understand.  

I have been using an ide that restarts interpreter sessions and therefore have not
run into this issue before.  I find it interesting, however, that I run into the same
problem when I restart the python interpreter (under interactive on the menu).  

I think briefly documenting this python idiosyncrasy a good idea... 
May I suggest a toolbar button that reloads the modules that have changed in the
editor?  (e.g., save-all-and-reload)  or perhaps something like the following
rollback code (which I happened across on-line, and can't say whether it might be
suitable) http://pyunit.sourceforge.net/notes/reloading.html Thanks, 
Derek

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-07T15:12:10Z_

Pierre, I completely agree with your comment about the behaviour of Python itself. 
Hoewever, I think Derek's last comment is worth noting further.  I experienced the
same behavior (changed modules not reloading) even after doing Restart Interpreter. 
The Restart generates a warning pop-up that seems to indicate that the workspace will
be totally reset including imported modules being dumped.  See attached.  It would
seem like the intended behavior of Reset Interpreter was to return the interpreter to
a completely clean state, but it doesn't seem to do that today.  It leaves some
things cached.  I can certainly see the value in having it work that way, and it
looks like that was probably the intent.

Pierre, this is a really amazing program..  I'm hooked.  Keep up the great work.

Thanks,

Jed

Version:  Spyder 1.0.0beta7
Platform:  XP

**Attachment:** [restart.png](http://code.google.com/p/spyderlib/issues/detail?id=62#hc3)

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-07T16:13:44Z_

After reading the Spyder documentation further, I realize now that the interactive
console is running in the same thread as the whole GUI.  So truly restarting the
interactive interpreter would essentially kill Spyder itself!  In that light, there
may not be a clean way to truly restart the interactive console.  Instead, maybe
consider updating the pop-up to warn the user that modules really aren't going to be
reloaded.  While specific bindings to modules might disappear, the modules themselves
aren't truly being reloaded when a restart happens.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-08T12:17:30Z_

Restarting the interactive console does unload the whole namespace -- i.e. all
references ("variables", functions, classes *and* modules). But, even if the
interactive console has been restarted, modules which were imported previously will
be imported using the same .pyc file as before, so changes in .py files won't be
taken into account.

Now, thanks to your excellent suggestion, I've added to 1.0.0rc2 (just released) the
"rollback importer" used in PyUnit. It seems to work very well but just in case it's
turned off by default (Source > Interactive console Settings > Force modules to be
reloaded).

Please do not hesitate to test it and report back in case anything goes wrong.

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [4.derek....@gmail.com](https://code.google.com/u/111899204237312999915/) on 2009-10-08T17:53:42Z_

Thanks Pierre, I will download and try it out.  I really like the software and I am
using it more and more... Doubtless, I will provide more feedback in future. 

Cheers, 
derek

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-08T22:59:54Z_

Well thank *you* for taking the time to report a bug, and please don't hesitate to do
it again in the near future as you've proposed.

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-09T10:20:16Z_

Pierre, here are some quick test results using the attached example code.  While it
seems to work right in some instances, there are some problems.

How to reproduce the problems:
1.  run test_reload.py in the interactive console.  A figure pops up.  You can close
the figure or leave it open, it doesn't matter.  
2.  Re-run test_reload.py in the interactive console.  Here is the output:

Traceback (most recent call last):
  File "C:\Python25\lib\site-packages\spyderlib\interpreter.py", line 82, in execfile
    self._install_rollback_importer()
  File "C:\Python25\lib\site-packages\spyderlib\interpreter.py", line 68, in
_install_rollback_importer
    self.rollback_importer.uninstall()
  File "C:\Python25\lib\site-packages\spyderlib\interpreter.py", line 38, in uninstall
    del sys.modules[modname]
KeyError: 'patches'

3.  Continue trying to re-run test_reload.py in interactive, and the same error
message occurs.
4.  From the UI, do "Restart Python interpreter."
5.  Re-run test_reload.py in interactive.  Same error message.
6.  Comment out line 11 (the plot() function) of test_reload.py and save.
7.  Exit Spyder and restart Spyder.
8.  Run and re-run test_reload.py in interactive, and everything seems to work now.
9.  Comment out line 9 of hello_world.py and save.
10.  Re-run test_reload.py interactive.  Seems to produce the desired result.  No
more "Hello, world."
11.  Uncomment line 9 of hello_world.py and save.
12.  From the UI, do "Restart Python interpreter."
13.  Re-run test_reload.py in interactive.  Seems to malfunction here.  No "Hello,
world." shows up but should have due to the code state of hello_world.py.
14.  Uncomment line 11 of test_reload.py, which re-activates the plot() function, and
save.
15.  Re-run test_reload.py in interactive, and the same error message as in step 2
appears.

Two separate problems appear to be at play here.  

First, there's something about the plot() function that causes things to fall apart
pretty severely.  The import of matplotlib doesn't seem to be the issue.  It's the
plot command itself.

Second, after a "Restart Python interpreter" command, the rollback importer seems to
stop reloading changed modules correclty.  Problems with the plot() function persist,
however, even after the interpreter restart.

Maybe there are too many complexities associated with this behavior that it may not
be worth the effort, but I'll leave that up to you!

Platform:  Sypder 1.0.0rc2 for Python 2.5 on XP

Regards,

Jed

**Attachment:** [hello_world.py test_reload.py](http://code.google.com/p/spyderlib/issues/detail?id=62#hc8)

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-10T01:17:08Z_

**Summary:** Interactive console won't recognize modified code in an imported module  
  **Status:** Started

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-10T07:31:42Z_

Good news: now it works!
Please test it in v1.0.0rc3 (or mercurial repository).

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-12T07:56:30Z_

One more round of test results for you.  The problems with plotting seem to be
resolved completely.  If you'll forgive my tenacity, there is one outstanding issue.
 The rollback importer ceases to function after doing a "Restart Python interpreter"
from the GUI.

Steps to reproduce:

1.  Use the same test files as in comment 8.  Run test_reload.py in interactive
console, and "Hello, world." is printed.
2.  Comment out line 9 of hello_world.py and re-run test_reload.py in interactive
console.  No printout.  Everything is working to this point.
3.  Uncomment line 9 of hello_world.py and save.
4.  Do "Restart Python interpreter" from the GUI menu.
5.  Re-run test_reload.py in interactive console.  No printout is displayed.  It
should be given the state of hello_world.py at that point.

Sorry for the bad news.  It's just about ready, but I think getting the final details
right is worth the effort.

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-12T07:58:23Z_

Forgot to say that the results in Comment 11 above were produced using Spyder
1.0.0rc3 in Python 2.5 on XP.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-12T09:28:33Z_

**Status:** Started

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-10-12T10:10:37Z_

fixed in v1.0.0rc4 (soon available as an .exe installer, now available as .tar.gz and
.egg)

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2009-10-12T11:13:13Z_

Tested.  Works beautifully now.  Thanks!

Config:  v1.0.0rc4 for Python 2.5 on XP
